### DISCUSSION TITLE
The listener for function 'Functions.CRONtrigger' was unable to start.

### DISCUSSION THREAD

**Discussion by RMHogervorst** [revodavid/R-custom-handler#6] [Q&A] [answered]
Maybe a bit out of scope for this specific case but, I'm trying something with a different trigger: the time trigger and I cannot get the docker container to work properly. 

For the timer trigger you need to connect a storage account or emulate the storage account with Azurite. (This all works on my mac using `func start`)
Building the docker image works
But when I  run the container it never starts the service, nor triggers. 

This is my dockerfile (I'm using the base container because I don't think we need dotnet, I tried a version using dotnet and there was no difference)

```
# To enable ssh & remote debugging on app service change the base image to the one below
#FROM mcr.microsoft.com/azure-functions/base:3.0-appservice
FROM mcr.microsoft.com/azure-functions/dotnet:3.0-appservice
ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
    AzureFunctionsJobHost__Logging__Console__IsEnabled=true

RUN apt update && \
    apt install -y  r-base
RUN apt-get update -qq && \
    apt-get install -y --no-install-recommends \
    libcurl4-openssl-dev \
    libsodium-dev \
    git-core \
    libssl-dev libxt6 && \
    mkdir -p ~/.local/share/renv && \
    R -e "install.packages('renv', repos='http://cran.rstudio.com/')"
### splitted the copying into parts so the rebuiliding times are quicker
COPY renv.lock /home/site/wwwroot/renv.lock
WORKDIR /home/site/wwwroot
RUN R -e "renv::restore(prompt=FALSE)"
COPY CRONtrigger/ /home/site/wwwroot/CRONtrigger
COPY host.json /home/site/wwwroot/host.json
COPY handler.R /home/site/wwwroot/handler.R
COPY handle-this.R /home/site/wwwroot/handle-this.R
```

The container log provides this help, I think it needs to access the storage emulator but never tells me what port it looks for, so I cannot open it on docker
Or should I copy my local.settings.json? that doesn't seem right either.

```
System.ArgumentNullException: Value cannot be null. (Parameter 'connectionString')
   at Microsoft.Azure.Storage.CloudStorageAccount.Parse(String connectionString)
   at Microsoft.Azure.WebJobs.Extensions.Timers.StorageScheduleMonitor.get_TimerStatusDirectory() in C:\azure-webjobs-sdk-extensions\src\WebJobs.Extensions\Extensions\Timers\Scheduling\StorageScheduleMonitor.cs:line 77
   at Microsoft.Azure.WebJobs.Extensions.Timers.StorageScheduleMonitor.GetStatusBlobReference(String timerName) in C:\azure-webjobs-sdk-extensions\src\WebJobs.Extensions\Extensions\Timers\Scheduling\StorageScheduleMonitor.cs:line 144
   at Microsoft.Azure.WebJobs.Extensions.Timers.StorageScheduleMonitor.GetStatusAsync(String timerName)
   at Microsoft.Azure.WebJobs.Extensions.Timers.Listeners.TimerListener.StartAsync(CancellationToken cancellationToken) in C:\azure-webjobs-sdk-extensions\src\WebJobs.Extensions\Extensions\Timers\Listener\TimerListener.cs:line 99
   at Microsoft.Azure.WebJobs.Host.Listeners.SingletonListener.StartAsync(CancellationToken cancellationToken) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Singleton\SingletonListener.cs:line 72
   at Microsoft.Azure.WebJobs.Host.Listeners.FunctionListener.StartAsync(CancellationToken cancellationToken, Boolean allowRetry) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Listeners\FunctionListener.cs:line 69
   --- End of inner exception stack trace ---
```

**Comment by RMHogervorst** [comment]
Underneath the strack trace it looks like this:

```
info: Host.Startup[413]
      Host started (375ms)
info: Host.Startup[0]
      Job host started
Hosting environment: Production
Content root path: /home/site/wwwroot
Now listening on: http://[::]:80
Application started. Press Ctrl+C to shut down.
fail: Host.Function.Console[0]
      Running plumber API at http://127.0.0.1:40293
fail: Host.Function.Console[0]
      Running swagger Docs at http://127.0.0.1:40293/__docs__/
info: Microsoft.Azure.WebJobs.Script.Workers.HttpFunctionInvocationDispatcher[0]
      Worker process started and initialized.
info: Host.Startup[0]
      Retrying to start listener for function 'Functions.CRONtrigger' (Attempt 1)
info: Host.Startup[0]
      Listener successfully started for function 'Functions.CRONtrigger' after 1 retries.
info: Host.General[316]
      Host lock lease acquired by instance ID '0000000000000000000000003CB61BEA'.
```

So the R function seems to start but the timer trigger never starts?

**Reply by revodavid** [reply]
Where are you running Docker and the container? Windows native, WSL2, or somewhere else?

**Reply by RMHogervorst** [reply]
I'm running docker on a mac

**Reply by RMHogervorst** [reply]
Maybe I should try it on a Linux computer so I can open all ports?

**Comment by revodavid** [comment]
If you can, it would be great if you could try it and let me know. There have been reports of port-related issues on Windows, but yours is the first on Mac. CC @anthonychu for visibility.

**Reply by RMHogervorst** [reply]
I created the same image on my ubuntu 20.04 laptop, and I ran the command 
`sudo docker run -it --network host <name of container>`    (using --network host in docker should allow docker to use the host network stack and so every connection would be opened) . Running on ubuntu leads to  the same error messages, so I would guess it is not a port issue. It seems I'm missing some sort of connectionString that is required for the listener function to start.

**Reply by anthonychu** [reply]
Timer triggers require an Azure Storage account. You need to provide a connection string in an environment variable named `AzureWebJobsStorage`. When running in Azure, there should already be a Storage account attached and the variable is set for you.

**Reply by RMHogervorst** [reply]
Yes, I had configured that in local.settings.json but how do I pass it to the docker container?  And even if I force the local settings in the container it gives me this error

**Reply by RMHogervorst** [reply]
I worked it out, you can pass an environmental variable to the container with -e but ultimately I think the easiest is with docker-compose (that way the network connections are automatic) 

So here is a working configuration for docker-compose with azurite for a timertrigger. 

```
version: "3.8"
services:
    storage:
        image: "mcr.microsoft.com/azure-storage/azurite"
    serverless:
        image: "YOURSERVERLESS_IMAGE:VERSION"
        environment:
            "AzureWebJobsStorage": "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://storage:10000/devstoreaccount1;"
```
These are the default credentials for azurite [as explained here]((https://github.com/azure/azurite#connection-strings).

**Comment by RMHogervorst** [comment]
solved here https://github.com/revodavid/R-custom-handler/discussions/6#discussioncomment-281849

**Comment by RMHogervorst** [comment, marked as answer]
I wrote the process down into a small blogpost here: https://blog.rmhogervorst.nl/blog/2021/01/18/testing-azure-functions-locally-with-azurite/
